homeassistant/components/sonos/media_player.py

```diff
@@ -134,6 +145,10 @@ def _discovered_player(soco):
             try:
                 _LOGGER.debug("Reached _discovered_player, soco=%s", soco)
 
+                # Avoid adding jobs when the event loop is closed
+                if hass.data[DATA_SONOS].stopped:
```

Comment:
Would it be possible to cancel the job that does the discovery instead? We keep a reference to that job and cancel it?

Is the pysonos discover thread already shutdown here and it's just our own job that risks doing work after event loop shutdown?

Reply:
The discovery thread is blocked in a `select()` for the duration of the discovery interval.

I am unsure how to cancel that so this seemed simpler :)

Reply:
The `event.call_later` call will return a function to remove the listener.

So there's no risk that the pysonos discover thread will spawn a discovery here? What endpoint is it waiting for with the select?

Reply:
That is a different case. The `event.call_later` is used for heartbeats with static configuration.

The discovery thread is an autonomous `threading.Thread` and looks like this:

```text
1 while not self.stopped:
2     if expired:
3         send discovery packet
4     response = select(timeout=60)
5     if response:
6         callback(response.player)
```

I don't know how to cancel line 4.

Reply:
Ok, so there's a risk of one callback before the thread knows that it should stop. Would it be safe to add a check on line 5, checking if the thread is stopped, before calling the callback?

Reply:
Well, no ... we could be preempted in the middle of line 6.

Even the current solution is not completely safe, it needs a lock on `hass.data[DATA_SONOS].stopped`.

Never mind, I will move the dirty bits to pysonos so it at least _looks_ neat. I will push an extra commit tonight.

Reply:
I think an alternative would be to use our dispatch helper in between the pysonos thread and our target callback. Then we could unsubscribe safely, on our side when needed.

Reply:
Or maybe not, the dispatcher still needs the event loop to schedule thread safely.

Reply:
I added the lock to pysonos so that should be good now (but not actually neat, I guess).

Regarding the static configuration: I now cancel the timer but will `block_till_done()` wait for the `event.call_later()` callback to complete if we are shutting down while it is already executing?

Reply:
`hass.async_block_till_done` will wait for tasks that are started after the stop event, eg the stop event listener.